File: modules/evtivity-attest/android/src/main/java/com/evtivity/attest/EvtivityAttestModule.kt
// Copyright (c) 2024-2026 EVtivity. All rights reserved.
// SPDX-License-Identifier: BUSL-1.1

package com.evtivity.attest

import com.google.android.play.core.integrity.IntegrityManagerFactory
import com.google.android.play.core.integrity.IntegrityTokenRequest
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

// Google Play Integrity bridge. The nonce passed from JS is the server
// challenge; the backend decodes the returned token and checks the nonce echoes
// back. Classic request flow: setCloudProjectNumber is optional for apps served
// by Play and only used when the caller supplies it.
class EvtivityAttestModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("EvtivityAttest")

    Function("isSupported") {
      true
    }

    AsyncFunction("requestIntegrityToken") { nonce: String, cloudProjectNumber: String?, promise: Promise ->
      val context = appContext.reactContext ?: run {
        promise.reject(Exceptions.ReactContextLost())
        return@AsyncFunction
      }
      val manager = IntegrityManagerFactory.create(context)
      val builder = IntegrityTokenRequest.builder().setNonce(nonce)
      cloudProjectNumber?.toLongOrNull()?.let { builder.setCloudProjectNumber(it) }
      manager
        .requestIntegrityToken(builder.build())
        .addOnSuccessListener { response -> promise.resolve(response.token()) }
        .addOnFailureListener { e -> promise.reject("INTEGRITY_FAILED", e.message ?: "Play Integrity request failed", e) }
    }

    AsyncFunction("generateKey") { promise: Promise ->
      promise.reject("UNSUPPORTED", "App Attest is iOS-only", null)
    }

    AsyncFunction("attestKey") { _: String, _: String, promise: Promise ->
      promise.reject("UNSUPPORTED", "App Attest is iOS-only", null)
    }

    AsyncFunction("generateAssertion") { _: String, _: String, promise: Promise ->
      promise.reject("UNSUPPORTED", "App Attest is iOS-only", null)
    }
  }
}

File: modules/evtivity-attest/ios/EvtivityAttestModule.swift
// Copyright (c) 2024-2026 EVtivity. All rights reserved.
// SPDX-License-Identifier: BUSL-1.1

import ExpoModulesCore
import DeviceCheck
import CryptoKit

// Apple App Attest bridge. The challenge passed from JS is the server nonce; the
// clientDataHash signed on device is SHA256(challenge), which the backend
// recomputes when verifying the attestation and each assertion.
public class EvtivityAttestModule: Module {
  public func definition() -> ModuleDefinition {
    Name("EvtivityAttest")

    Function("isSupported") { () -> Bool in
      if #available(iOS 14.0, *) {
        return DCAppAttestService.shared.isSupported
      }
      return false
    }

    AsyncFunction("generateKey") { (promise: Promise) in
      guard #available(iOS 14.0, *), DCAppAttestService.shared.isSupported else {
        promise.reject("UNSUPPORTED", "App Attest is not supported on this device")
        return
      }
      DCAppAttestService.shared.generateKey { keyId, error in
        if let error = error {
          promise.reject("KEYGEN_FAILED", error.localizedDescription)
          return
        }
        guard let keyId = keyId else {
          promise.reject("KEYGEN_FAILED", "generateKey returned no key id")
          return
        }
        promise.resolve(keyId)
      }
    }

    AsyncFunction("attestKey") { (keyId: String, challenge: String, promise: Promise) in
      guard #available(iOS 14.0, *) else {
        promise.reject("UNSUPPORTED", "App Attest is not supported on this device")
        return
      }
      let clientDataHash = Data(SHA256.hash(data: Data(challenge.utf8)))
      DCAppAttestService.shared.attestKey(keyId, clientDataHash: clientDataHash) { attestation, error in
        if let error = error {
          promise.reject("ATTEST_FAILED", error.localizedDescription)
          return
        }
        guard let attestation = attestation else {
          promise.reject("ATTEST_FAILED", "attestKey returned no object")
          return
        }
        promise.resolve(attestation.base64EncodedString())
      }
    }

    AsyncFunction("generateAssertion") { (keyId: String, challenge: String, promise: Promise) in
      guard #available(iOS 14.0, *) else {
        promise.reject("UNSUPPORTED", "App Attest is not supported on this device")
        return
      }
      let clientDataHash = Data(SHA256.hash(data: Data(challenge.utf8)))
      DCAppAttestService.shared.generateAssertion(keyId, clientDataHash: clientDataHash) { assertion, error in
        if let error = error {
          promise.reject("ASSERT_FAILED", error.localizedDescription)
          return
        }
        guard let assertion = assertion else {
          promise.reject("ASSERT_FAILED", "generateAssertion returned no object")
          return
        }
        promise.resolve(assertion.base64EncodedString())
      }
    }

    AsyncFunction("requestIntegrityToken") { (_ nonce: String, _ cloudProjectNumber: String?, promise: Promise) in
      promise.reject("UNSUPPORTED", "Play Integrity is Android-only")
    }
  }
}
